### PR TITLE
CE-3552, Open Infobox builder from VE Context

### PR DESCRIPTION
**Pull request by RafLeszczynski**
## Links

* https://wikia-inc.atlassian.net/browse/CE-3552
* [Related dev changes #app/pull/10150](https://github.com/Wikia/app/pull/10150)

## Description

Open Infobox builder from VE Context
## Reviewers

@Grunny @adamkarminski @lukaszkonieczny @Wikia/x-wing @dianafa

### CODE REVIEW

File: front/main/app/controllers/infobox-builder.js [hunk 1]
```diff
@@ -13,7 +25,18 @@ export default Ember.Controller.extend({
...
 		cancel() {
-			this.get('target').send('redirectToPage');
+			if (this.get('isVEContext')) {
+				if (!this.get('isDirty') || window.confirm(this.confirmationMessage())) {
+					const model = this.get('model');
+
+					this.get('target').send('returnToVE');
+					this.set('isDirty', false);
+					model.initInfoboxState();
+					model.setupInitialState();
```
Comment (dianafa) [inline front/main/app/controllers/infobox-builder.js]: what about moving above lines to a separate function?

File: front/main/app/controllers/infobox-builder.js [hunk 2]
```diff
@@ -47,6 +70,15 @@ export default Ember.Controller.extend({
 			return this.get('model').getTemplateExists(title);
 		},
 
+		goBackToVE() {
+			const model = this.get('model'),
+				route = this.get('target');
+
+			route.send('returnToVE', true);
+			model.initInfoboxState();
+			model.setupInitialState();
```
Comment (dianafa) [inline front/main/app/controllers/infobox-builder.js]: Why do we need to cleanup infobox state before (after :P ) leaving? `returnToVE` is a promise, doesn't it matter if we run initInfoboxState & setupInitialState before or after it resolves?

File: front/main/app/components/infobox-builder.js [hunk 3]
```diff
@@ -384,15 +384,17 @@ export default Ember.Component.extend(
 		 */
 		handleSaveResults(data, shouldRedirectToPage) {
```
Comment (RafLeszczynski) [inline front/main/app/components/infobox-builder.js]: The JsDoc description for this method is no longer valid after adding VE integration. Please update the JsDoc

File: front/main/app/components/infobox-builder.js [hunk 4]
```diff
@@ -384,15 +384,17 @@ export default Ember.Component.extend(
...
 			if (data.success) {
-				this.set('showSuccess', true);
+				this.set('showSuccess', !this.get('isVEContext'));
```
Comment (RafLeszczynski) [inline front/main/app/components/infobox-builder.js]: why you don't want to show success after save ?

File: front/main/app/components/infobox-builder.js [hunk 5]
```diff
@@ -384,15 +384,17 @@ export default Ember.Component.extend(
...
 
 				track({
 					action: trackActions.success,
 					category: 'infobox-builder',
 					label: 'save-successful'
 				});
 
-				if (shouldRedirectToPage) {
+				if (this.get('isVEContext')) {
```
Comment (RafLeszczynski) [inline front/main/app/components/infobox-builder.js]: please consider caching isVEContext in local const as it's used multiple times in this method

File: front/main/app/controllers/infobox-builder.js [hunk 6]
```diff
@@ -47,6 +70,15 @@ export default Ember.Controller.extend({
 			return this.get('model').getTemplateExists(title);
 		},
 
+		goBackToVE() {
```
Comment (RafLeszczynski) [inline front/main/app/controllers/infobox-builder.js]: missing JS doc

File: front/main/app/controllers/infobox-builder.js [hunk 7]
```diff
@@ -13,7 +25,18 @@ export default Ember.Controller.extend({
 		 * @returns {void}
 		 */
 		cancel() {
-			this.get('target').send('redirectToPage');
+			if (this.get('isVEContext')) {
+				if (!this.get('isDirty') || window.confirm(this.confirmationMessage())) {
```
Comment (RafLeszczynski) [inline front/main/app/controllers/infobox-builder.js]: As controllers are more on less deprecated in Ember and they will be soon removed in considered a good pattern to make them as light as possible. Usually we use them just for  handling actions and calling appropriate methods on route or doing some operations required for changing UI that cannot be done on component level. 

I'd propose moving this logic to separate method on route. 

In addition please try to avoid working on a single model instance for multiple infobox templates and resetting the model after successful save. AFAIK you need to have clear model state each time you are lunching IB from within VE. Instead doing operations on single model instance please create new instance each time you open IB in VE and pass this instance to controller.

File: front/main/app/models/infobox-builder.js [hunk 8]
```diff
@@ -7,14 +7,20 @@ const InfoboxBuilderModel = Ember.Object.extend({
...
-			image: 0,
-			title: 0,
-			'section-header': 0
-		};
-		this.infoboxState = [];
-		this.itemInEditMode = null;
+		this.initInfoboxState();
+	},
+
+	initInfoboxState() {
```
Comment (RafLeszczynski) [inline front/main/app/models/infobox-builder.js]: Please check the comment above. instead of resetting model instance to initial sate consider simply creating new model instance and pass it to controller

File: front/main/app/routes/infobox-builder.js [hunk 9]
```diff
@@ -312,5 +324,19 @@ export default Ember.Route.extend(ConfirmationMixin, {
...
+	 * set VE context - true if IB opened inside visual editor
+	 * @param {Boolean} isVEContext
+	 * @returns {void}
+	 */
+	setVEContext(isVEContext = false) {
+		const controller = this.controllerFor('infobox-builder');
+
+		controller.set('isVEContext', isVEContext);
+		if (isVEContext) {
+			controller.set('isDirty', false);
```
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: why do you set `isDirty` to `false` for  `isVEContext`?

**Comment by RafLeszczynski** [comment]
Only part of new introduced logic is covered by tests, please add tests for remaining part.

File: front/main/app/models/infobox-builder.js (hunk 8 above)
Comment (RafLeszczynski) [inline front/main/app/models/infobox-builder.js]: Please consider triggering `refresh` on route: http://emberjs.com/api/classes/Ember.Route.html#method_refresh

You will need to refactor `beforeModel` hook in `infobox-builder.js` route so that ponto in not loaded second time and styles are not added to DOM also for the second time.

File: front/main/app/components/infobox-builder.js (hunk 4 above)
Comment (bognix) [inline front/main/app/components/infobox-builder.js]: is it because ponto iframe is closed and message about success is not shown?

File: front/main/app/components/infobox-builder.js (hunk 5 above)
Comment (lukaszkonieczny) [inline front/main/app/components/infobox-builder.js]: Not anymore ;)

File: front/main/app/routes/infobox-builder.js (hunk 9 above)
Comment (lukaszkonieczny) [inline front/main/app/routes/infobox-builder.js]: Not to show confirmation dialog when there is no change made by user

File: front/main/app/controllers/infobox-builder.js [hunk 10]
```diff
@@ -13,7 +14,11 @@ export default Ember.Controller.extend({
 		 * @returns {void}
 		 */
 		cancel() {
-			this.get('target').send('redirectToPage');
+			if (this.get('isVEContext')) {
+				this.get('target').send('returnToVE');
```
Comment (RafLeszczynski) [inline front/main/app/controllers/infobox-builder.js]: to avoid code duplication you can just call `goBackToVE` action using `this.send('goBackToVE')`

File: front/main/app/routes/infobox-builder.js (hunk 9 above)
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: but if you create a new template it already has a default items added to the stack so there is a change and user will lost what's in a preview if she exits IB without save

File: front/main/app/routes/infobox-builder.js [hunk 11]
```diff
@@ -207,9 +229,12 @@ export default Ember.Route.extend(ConfirmationMixin, {
 	 * @returns {JQueryXHR}
```
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: return type is not valid after code changes

File: front/main/app/routes/infobox-builder.js [hunk 12]
```diff
@@ -207,9 +229,12 @@ export default Ember.Route.extend(ConfirmationMixin, {
...
 	 */
 	loadPonto() {
```
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: please consider changing this method name as it currently loads ponto only if it's not already being loaded or what what would be even better move the conditional to the `beforeModel` hook to keep this method responsible only for single action

File: front/main/app/routes/infobox-builder.js [hunk 13]
```diff
@@ -28,7 +28,7 @@ export default Ember.Route.extend(ConfirmationMixin, {
 						this.setupStyles(responses.dataAndAssets);
```
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: `setupStyles` will be called each time user reopens dialog with IB in VE which will cause multiplication of CSS assets added to DOM. It should be called only ones for initial load of Mercury similar to Ponto

File: front/main/app/routes/infobox-builder.js [hunk 14]
```diff
@@ -28,7 +28,7 @@ export default Ember.Route.extend(ConfirmationMixin, {
...
 						this.setupInfoboxData(responses.dataAndAssets);
 					})
-					.then(this.isWikiaContext)
+					.then(this.isWikiaContext.bind(this))
```
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: do we want to check context each time we reload model for this route or should it be done only ones as context don't change?

File: front/main/tests/unit/controllers/infobox-builder-test.js [hunk 15]
```diff
@@ -137,3 +137,49 @@ test('test group items calculation', function (assert) {
 		assert.deepEqual(controller.get('groupItems'), testCase.expected);
 	});
 });
+
+test('check the appropriate action is sent on save and cancel', function (assert) {
+	const controller = this.subject(),
+		model = {
+			saveStateToTemplate: sinon.stub().returns(Ember.RSVP.Promise.resolve()),
+			initInfoboxState: sinon.stub(),
```
Comment (RafLeszczynski) [inline front/main/tests/unit/controllers/infobox-builder-test.js]: `initInfoboxState` method is not declared in `infobox-builder` model

**Comment by RafLeszczynski** [comment]
please fix failing tests

File: front/main/tests/unit/controllers/infobox-builder-test.js [hunk 16]
```diff
@@ -137,3 +137,49 @@ test('test group items calculation', function (assert) {
...
+				isVEContext: true,
+				calledWith: 'returnToVE'
+			},
+			{
+				action: 'cancel',
+				isVEContext: false,
+				calledWith: 'redirectToPage'
+			},
+			{
+				action: 'save',
```
Comment (RafLeszczynski) [inline front/main/tests/unit/controllers/infobox-builder-test.js]: `save` action doesn't check `isVEContext` any more so these tests should be refactored

File: front/main/app/components/infobox-builder.js (hunk 5 above)
Comment (RafLeszczynski) [inline front/main/app/components/infobox-builder.js]: please update `handleSaveResults` - https://github.com/Wikia/mercury/blob/CE-3552/front/main/tests/unit/components/infobox-builder-test.js#L503

File: front/main/app/routes/infobox-builder.js [hunk 17]
```diff
@@ -1,10 +1,10 @@
 import Ember from 'ember';
-import {track, trackActions} from 'common/utils/track';
 import InfoboxBuilderModel from '../models/infobox-builder';
 import ConfirmationMixin from 'ember-onbeforeunload/mixins/confirmation';
+import {track, trackActions} from 'common/utils/track';
 
 export default Ember.Route.extend(ConfirmationMixin, {
-	pontoLoadingInitialized: false,
+	modelInitialized: false,
```
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: you are using this flag to check if `ponto` should be loaded, assets added to DOM and  `wikiContext` set. It doesn't have to do anything with initialising the model so consider changing the name for this flag. Maybe `isEnvirionmentSet` or something else???

File: front/main/app/routes/infobox-builder.js [hunk 18]
```diff
@@ -17,18 +17,24 @@ export default Ember.Route.extend(ConfirmationMixin, {
...
 		return new Ember.RSVP.Promise((resolve, reject) => {
-			if (window.self !== window.top && (!window.Ponto || !this.get('pontoLoadingInitialized'))) {
+			if (window.self !== window.top) {
 				const promises = {
-					dataAndAssets: this.loadInfoboxDataAndAssets(templateName),
-					ponto: this.loadPonto()
+					dataAndAssets: this.loadInfoboxDataAndAssets(templateName)
 				};
 
+				if (!window.Ponto || !this.get('modelInitialized')) {
```
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: please consider moving this condition into computed property `shouldLoadPonto`

File: front/main/app/routes/infobox-builder.js [hunk 19]
```diff
@@ -17,18 +17,24 @@ export default Ember.Route.extend(ConfirmationMixin, {
 		const templateName = transition.params['infobox-builder'].templateName;
 
 		return new Ember.RSVP.Promise((resolve, reject) => {
-			if (window.self !== window.top && (!window.Ponto || !this.get('pontoLoadingInitialized'))) {
+			if (window.self !== window.top) {
```
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: you could also create a computed property `isIframeContext` and move this check there

File: front/main/app/routes/infobox-builder.js [hunk 20]
```diff
@@ -17,18 +17,24 @@ export default Ember.Route.extend(ConfirmationMixin, {
...
+				if (!window.Ponto || !this.get('modelInitialized')) {
+					promises.ponto = this.loadPonto();
+				}
+
 				Ember.RSVP.hash(promises)
 					.then((responses) => {
-						this.setupStyles(responses.dataAndAssets);
+						if (!this.get('modelInitialized')) {
+							this.setupStyles(responses.dataAndAssets);
+							this.isWikiaContext();
```
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: `isWikiaContext` returns promise, are you sure that the whole promise chain will work properly if `isWikiaContext` will be rejected?

File: front/main/tests/unit/controllers/infobox-builder-test.js [hunk 21]
```diff
@@ -137,3 +137,37 @@ test('test group items calculation', function (assert) {
 		assert.deepEqual(controller.get('groupItems'), testCase.expected);
 	});
 });
+
+test('check the appropriate action is sent on save and cancel', function (assert) {
```
Comment (RafLeszczynski) [inline front/main/tests/unit/controllers/infobox-builder-test.js]: the name of the test doesn't correspond with the test itself - `save` action is not being tested here

File: front/main/app/routes/infobox-builder.js [hunk 22]
```diff
@@ -17,18 +17,24 @@ export default Ember.Route.extend(ConfirmationMixin, {
 		const templateName = transition.params['infobox-builder'].templateName;
```
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: `beforeModel` has become pretty complex with many nested `if` statements. Please consider covering this method with tests so we are sure that all cases work as expected. In addition it would be also great if you think about reducing `if` and / or breaking this into smaller methods that would easier to test. 

Additionally consider moving conditions into descriptive computed properties so its much easier to understand what's going on  - check comments below

File: front/main/app/controllers/infobox-builder.js [hunk 23]
```diff
@@ -13,7 +14,11 @@ export default Ember.Controller.extend({
 		 * @returns {void}
 		 */
 		cancel() {
-			this.get('target').send('redirectToPage');
+			if (this.get('isVEContext')) {
+				this.send('goBackToVE');
```
Comment (SebastianMarzjan) [inline front/main/app/controllers/infobox-builder.js]: Why is this needed?
could https://github.com/Wikia/mercury/pull/2280/files#diff-e18b400a25ffa404ccfd4fa8ca7fe5b6R61 be invoked directly here?

File: front/main/tests/unit/controllers/infobox-builder-test.js (hunk 21 above)
Comment (RafLeszczynski) [inline front/main/tests/unit/controllers/infobox-builder-test.js]: please address comment above

File: front/main/app/templates/components/infobox-builder.hbs [hunk 24]
```diff
@@ -72,17 +72,19 @@
...
-			click=(action 'tryGoToSource')
-			icon='source'
-			iconSize=24
-			id='go-to-source'
-		}}
-		{{#pop-over for="go-to-source" on="hover" flow='leftCenter'}}
-			<p class="infobox-builder-go-to-source-tooltip-text">{{i18n 'main.source-editor-tooltip'
-			ns='infobox-builder'}}</p>
-		{{/pop-over}}
+		{{#unless isVEContext}}
```
Comment (RafLeszczynski) [inline front/main/app/templates/components/infobox-builder.hbs]: could we change this variable to `isGoToSoruceEnabled` or something similar. IB should now anything about its communication with VE on template basis

File: front/main/app/routes/infobox-builder.js [hunk 25]
```diff
@@ -141,6 +164,43 @@ export default Ember.Route.extend(ConfirmationMixin, {
...
 
 	/**
+	 * Setup infobox builder by loading infobox data and styles.
+	 * Also initialize ponto and checks in what context
+	 * infobox builder was opened
+	 *
+	 * @param {string} templateName
+	 * @returns {Promise}
+	 */
+	setupEnvironment(templateName) {
```
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: what do you think about `setupEnvironmentAndInfoboxData` ?

File: front/main/app/routes/infobox-builder.js [hunk 26]
```diff
@@ -141,6 +164,43 @@ export default Ember.Route.extend(ConfirmationMixin, {
...
+
+	/**
+	 * Setup infobox data.
+	 * It's invoke on model refresh
+	 * to avoid loading all existing resources again
+	 *
+	 * @param {string} templateName
+	 * @returns {Promise}
+	 */
+	updateInfoboxData(templateName) {
```
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: what do you think about `loadAndSetupInfoboxData` ?

File: front/main/app/routes/infobox-builder.js [hunk 27]
```diff
@@ -17,20 +20,19 @@ export default Ember.Route.extend(ConfirmationMixin, {
...
-					.then((responses) => {
-						this.setupStyles(responses.dataAndAssets);
-						this.setupInfoboxData(responses.dataAndAssets);
-					})
-					.then(this.isWikiaContext)
-					.then(resolve)
-					.catch(reject);
+			if (this.get('isIframeContext')) {
+				if (!this.get('isEnvironmentSet')) {
+					this.setupEnvironment(templateName)
```
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: I'd suggest doing the following:
- break getting assets and getting data into two separate requests so we do not need to request the same things twice as it being done right now.
- this will allow us for first setting the environment which is Ponto, assets and wikia context and later request for template data and create a model from it
- in addition I'd think about moving ponto communication and assets into separate `Ember.Service` or maybe more than one service

There is the question when and how this work should be done. As I am WFH today please talk to @SebastianMarzjan  about how we can handle that

File: front/main/app/controllers/infobox-builder.js [hunk 28]
```diff
@@ -47,6 +52,15 @@ export default Ember.Controller.extend({
 			return this.get('model').getTemplateExists(title);
 		},
 
+		/**
+		 * Triggers action to communicate VE that creating infobox is completed
+		 *
+		 * @returns {void}
+         */
+		goBackToVE() {
```
Comment (RafLeszczynski) [inline front/main/app/controllers/infobox-builder.js]: could you make action names consistent - either `goBackToVE` or `returnToVE`

File: front/main/app/routes/infobox-builder.js [hunk 29]
```diff
@@ -17,20 +20,19 @@ export default Ember.Route.extend(ConfirmationMixin, {
...
-
-				Ember.RSVP.hash(promises)
-					.then((responses) => {
-						this.setupStyles(responses.dataAndAssets);
-						this.setupInfoboxData(responses.dataAndAssets);
-					})
-					.then(this.isWikiaContext)
-					.then(resolve)
-					.catch(reject);
+			if (this.get('isIframeContext')) {
```
Comment (RafLeszczynski) [inline front/main/app/routes/infobox-builder.js]: this code is also a great candidate for being covered with tests :-)

File: front/main/app/routes/infobox-builder.js (hunk 27 above)
Comment (kamilkoterba) [inline front/main/app/routes/infobox-builder.js]: Added TODO
https://github.com/Wikia/mercury/pull/2280/commits/9d83f6e464cc222bb3069818e76e569ba94e92fa
https://wikia-inc.atlassian.net/browse/CE-3600

File: front/main/app/routes/infobox-builder.js (hunk 12 above)
Comment (kamilkoterba) [inline front/main/app/routes/infobox-builder.js]: Done https://github.com/Wikia/mercury/pull/2280/commits/35e9199896e88f05f1815b24611c1e0070a95126

File: front/main/app/components/infobox-builder.js [hunk 30]
```diff
@@ -393,8 +393,8 @@ export default Ember.Component.extend(
 					label: 'save-successful'
 				});
 
-				if (this.get('isVEContext')) {
-					this.get('goBackToVE')();
+				if (this.get('isGoToSourceEnabled')) {
```
Comment (RafLeszczynski) [inline front/main/app/components/infobox-builder.js]: ? is this intended?? why do want to check `isGoToSourceEnabled` here ?

File: front/main/app/templates/components/infobox-builder.hbs [hunk 31]
```diff
@@ -72,7 +72,7 @@
 				}}
 			</div>
 		{{/liquid-bind}}
-		{{#unless isVEContext}}
+		{{#unless isGoToSourceEnabled}}
```
Comment (RafLeszczynski) [inline front/main/app/templates/components/infobox-builder.hbs]: something is wrong with this condition…???? unless `isGoToSourceEnabled === true` render go to source button ???? this means that go to source button should be rendered when `isGoToSourceEnabled === false`

File: front/main/app/routes/infobox-builder.js (hunk 11 above)
Comment (kamilkoterba) [inline front/main/app/routes/infobox-builder.js]: It's valid now

File: front/main/app/components/infobox-builder.js (hunk 30 above)
Comment (kamilkoterba) [inline front/main/app/components/infobox-builder.js]: This part is related to isVEContext check. I think that this redirect/goBack to VE logic should be moved to controller...

File: front/main/app/components/infobox-builder.js [hunk 32]
```diff
@@ -392,7 +393,10 @@ export default Ember.Component.extend(
 					label: 'save-successful'
 				});
 
-				if (shouldRedirectToPage) {
+				if (!this.get('isGoToSourceEnabled')) {
```
Comment (RafLeszczynski) [inline front/main/app/components/infobox-builder.js]: why are we checking the value of `isGoToSourceEnabled` to decide if we want to get back to VE?

File: front/main/app/controllers/infobox-builder.js [hunk 33]
```diff
@@ -5,6 +5,8 @@ export default Ember.Controller.extend({
 	isDirty: false,
 	groupItems: [],
 	lastGroupItem: null,
+	isVEContext: false,
+	isGoToSourceEnabled: Ember.computed.not('isVEContext'),
```
Comment (RafLeszczynski) [inline front/main/app/controllers/infobox-builder.js]: consider moving this computed property to component as you still need `isVEContext` on component level to trigger the right action - check the comment above

File: front/main/app/components/infobox-builder.js (hunk 32 above)
Comment (kamilkoterba) [inline front/main/app/components/infobox-builder.js]: As I wrote above:

This part is related to isVEContext check and I think it doesn't belong here. I think that this redirect/goBack to VE logic should be moved to controller. Then we'll get rid of VE related functionality within this component, which would make it agnostic to VE context.

I'm working on it.

File: front/main/app/controllers/infobox-builder.js (hunk 10 above)
Comment (kamilkoterba) [inline front/main/app/controllers/infobox-builder.js]: per @SebastianMarzjan comment below decided to call directly

File: front/main/app/routes/infobox-builder.js (hunk 22 above)
Comment (kamilkoterba) [inline front/main/app/routes/infobox-builder.js]: Addressed https://github.com/Wikia/mercury/pull/2280/commits/e8a6ac2aab75b378ab7ef2c6ae208ed25a4a31db

**Comment by RafLeszczynski** [comment]
LGTM if build pass
